src/test/java/io/pinecone/BuildUpsertVectorWithUnsignedIndicesTest.java

```diff
@@ -0,0 +1,110 @@
+import io.pinecone.unsigned_indices_model.VectorWithUnsignedIndices;
+import org.junit.jupiter.api.Test;
+
+import static io.pinecone.GenerateMetadata.generateMetadataStruct;
+import static io.pinecone.commons.IndexInterface.buildUpsertVectorWithUnsignedIndices;
+import static org.junit.jupiter.api.Assertions.*;
+
+import java.util.Arrays;
+import java.util.List;
+
+public class BuildUpsertVectorWithUnsignedIndicesTest {
+
+    @Test
+    void testRequiredAndOptionalParameters() {
+        String id = "some_id";
+        List<Float> values = Arrays.asList(1.0f, 2.0f, 3.0f);
+        List<Long> sparseIndices = Arrays.asList(0L, 2L);
+        List<Float> sparseValues = Arrays.asList(1.5f, 3.5f);
+        SparseValuesWithUnsignedIndices sparseValuesWithUnsignedIndices = new SparseValuesWithUnsignedIndices(sparseIndices, sparseValues);
+        Struct metadata = generateMetadataStruct(0, 0);
+
+        VectorWithUnsignedIndices result = new VectorWithUnsignedIndices(id, values, metadata, sparseValuesWithUnsignedIndices);
+
+        assertNotNull(result);
+        assertEquals(id, result.getId());
+        assertEquals(values, result.getValuesList());
+        assertNotNull(result.getSparseValuesWithUnsignedIndices());
+        assertEquals(sparseIndices, result.getSparseValuesWithUnsignedIndices().getIndicesWithUnsigned32IntList());
+        assertEquals(sparseValues, result.getSparseValuesWithUnsignedIndices().getValuesList());
+        assertEquals(metadata, result.getMetadata());
+    }
+
+    @Test
+    void testRequiredParametersOnly() {
+        String id = "some_id";
+        List<Float> values = Arrays.asList(1.0f, 2.0f, 3.0f);
+        SparseValuesWithUnsignedIndices sparseValuesWithUnsignedIndices = null;
+        Struct metadata = null;
+
+        VectorWithUnsignedIndices result = new VectorWithUnsignedIndices(id, values, metadata, sparseValuesWithUnsignedIndices);
+
+        assertNotNull(result);
+        assertEquals(id, result.getId());
+        assertEquals(values, result.getValuesList());
+        assertNull(result.getSparseValuesWithUnsignedIndices());
+        assertNull(result.getMetadata());
+    }
+
+    @Test
+    void testNullId() {
+        String id = null;
+        List<Float> values = Arrays.asList(1.0f, 2.0f, 3.0f);
+        List<Long> sparseIndices = Arrays.asList(0L, 2L);
+        List<Float> sparseValues = Arrays.asList(1.5f, 3.5f);
+        Struct metadata = Struct.newBuilder().build();
+        StringBuilder exceptionMessage = new StringBuilder();
+
+        try {
+            buildUpsertVectorWithUnsignedIndices(id, values, sparseIndices, sparseValues, metadata);
+        } catch (PineconeValidationException exception) {
+            exceptionMessage.append(exception.getLocalizedMessage());
+        } finally {
+            assertEquals(exceptionMessage.toString(), "Invalid upsert request. Please ensure that both id " +
+                    "and values are provided.");
+        }
```

Comment:
Do we need a `catch` / `finally` for these? I don't really get why we need `exceptionMessage` at all, can we not just `assertEquals(exception.getLocalizedMessage(), "Invalid upsert request. Please ensure that both id " +
                    "and values are provided.");` directly?

Reply:
If there's no exception or some other exception occurs, the catch (PineconeValidationException exception) will not be executed. Finally block now ensures that the expected exception occurs.
https://github.com/pinecone-io/pinecone-java-client/pull/80#discussion_r1523452274